src/gt4sd/frameworks/enzeptional/genetic_algorithm.py

```diff
@@ -0,0 +1,229 @@
+        number_of_sequences: Optional[int] = None,
+        seed: int = 42,
+        time_budget: Optional[int] = None,
+        mutations: Mutations = Mutations(IUPAC_MUTATION_MAPPING),
+    ) -> List[Dict[str, Any]]:
+        """Optimize the enzyme given a number of mutations and a range.
+
+        If the range limits are not provided the full sequence is optimized, this might be inefficient.
+        The sampling is performing by exploring mutations with a slightly smart random sampling.
+
+        Args:
+            number_of_mutations: number of allowed mutations.
+            intervals: list of ranges in the sequence, zero-based. Defaults to None, a.k.a. use optimize the full sequence.
+            number_of_steps: number of optimization steps. Defaults to 100.
+            number_of_samples_per_step: number of samples sequences per optimization step. Defaults to 32.
+            number_of_sequences: number of optimal seuqence returned. Defaults to None, a.k.a, returns all.
+            seed: seed for random number generation. Defaults to 42.
+            time_budget: maximum allowed runtime in seconds. Defaults to None, a.k.a., no time limit, running for number_of_steps steps.
+            mutations: mutations definition. Defaults to uniform sampling of IUPAC AAs.
+
+        Raises:
+            ValueError: in case an invalid range is provided.
+
+        Returns:
+            a list of dictionaries containing a candidate optimal sequence and the related score. Sorted from best to worst.
+            Note that, when no limit on the returned number of sequences is set, the worst sequence is the original unmutated sequence.
+            If the optimization fails, only the original sequence is returned.
+        """
+        random.seed(seed)
+
+        # check if interval is None. In case it is, take as interval the whole sequence
+        if intervals is None:
+            intervals = [(0, self.sequence_length)]
+        else:
+            intervals = sanitize_intervals(
+                intervals
+            )  # here we merged and sorted the intervals
+
+        # check that the intervals are in the range of the sequence length
+        if intervals[-1][1] > self.sequence_length:
+            raise ValueError(
+                "check provided intervals, at least an interval is larger than the sequence length"
+            )
+
+        # create a sequence from based on the intervals
+        sequence_from_intervals = "".join(
+            [self.sequence[start:end] for start, end in intervals]
+        )
+
+        # mutate the sequence from intervals
+        aa_sequence_range = AASequence(sequence_from_intervals, mutations=mutations)
+        maximum_number_of_mutations = number_of_mutations
+        logger.info(
+            f"maximum number of mutations for the intervals: {maximum_number_of_mutations}"
+        )
+        if self.protein_embedding == "ProtTransXL" in str(type(self.protein_embedding)):
```

Comment:
this need to be changed, `self.protein_embedding` should be a `StringEmbedding` class, then everything should become consistent.